design/design-doc-resource-claim-definition.md

```diff
@@ -0,0 +1,158 @@
+# Resource Claim Definition
+* Owner: Nic Cope (@negz)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Background
+
+See the following issues this design doc intends to address:
+
+* https://github.com/crossplaneio/crossplane/issues/1105
+* https://github.com/crossplaneio/crossplane/issues/1106
+* https://github.com/crossplaneio/crossplane/issues/1107
+
+TODO(negz): Write this up.
+
+## Goals
+
+* Allow infrastructure operators to define claim kinds.
+* Allow a claim to be satisfied by multiple resources.
+* Surface claim to resource compatibility in Crossplane's API.
+
+TODO(negz): Write this up.
+
+## Proposal
+
+This document proposes the introduction of a new, cluster scoped, resource
+named `ResourceClaimDefinition`. A `ResourceClaimDefinition` builds on the
+existing `CustomResourceDefinition` concept, allowing an infrastructure
+operator to define arbitrary kinds of resource claim. Operators may define the
+schema for their claim, the resources it may bind to, and how it may dynamically
+provision those resources all without needing to write any code.
+
+```yaml
+apiVersion: core.crossplane.io/v1alpha1
+kind: ResourceClaimDefinition
+metadata:
+  name: mysqlinstance
+spec:
+  # A controller watching for ResourceClaimDefinition uses the group, names,
+  # version, and validation fields to create a CustomResourceDefinition.
```

Comment:
@bassam had similar feedback. I was using a similar approach at first in which this resource would reference a claim CRD rather than define it. I ended up merging them because I feel (quite strongly) that we need a layer of abstraction atop "write a CRD freehand" for claim authors, so I put that in here.

However I could see it being more flexible if we took something like this and broke it up into two kinds, e.g.:

* `ResourceClaimDefinition` - Makes it easy to write compliant and sensible resource claim CRDs. Basically a subset of `CustomResourceDefinition`, each produces a CRD.
* `ResourceClaimSatsifaction` - Basically everything covered by a "resource group" in this sketch.

Reply:
I've pushed a change that breaks this into two kinds.